### PR TITLE
Feature: Communication Matrix

### PR DESCRIPTION
**Pull request by adityaranjan**
- basic outline for tracedata with definitions and events
- p2p comm matrix function
- plan to add list of roots and calling context graph as well

### CODE REVIEW

**Comment by jordanmarry** [comment]
If we wanted to use this for nsight/hpctoolkit readers, wouldn't there be issues since those two files don't provide a Location Group ID column?

**Comment by adityaranjan** [comment]
> If we wanted to use this for nsight/hpctoolkit readers, wouldn't there be issues since those two files don't provide a Location Group ID column?

**Comment by adityaranjan** [comment]
> If we wanted to use this for nsight/hpctoolkit readers, wouldn't there be issues since those two files don't provide a Location Group ID column?

From what Alex has mentioned about hpctookit, it does not provide any sender/receiver or message size information for MPI calls, so this function can not be used for that reader. I am not sure about NSight. I don't think we have enough data there to know what kind of information for MPI calls are traced with the MPI wrapper for NVTX. If you find any such data, let me know and we can make it work.

The column name differences does bring up a more general concern, which I have discussed in this comment of yours:
https://github.com/hpcgroup/pipit/pull/19#issuecomment-1243233166

**Review by jordanmarry — APPROVED** [review]

**Comment by adityaranjan** [comment]
To Do:
Check on the time taken for this function and possibly optimize.
See if there is any easy way to parallelize.

**Comment by hsirkar** [comment]
Which dimensions are the senders/receivers?

For example in this matrix:
```
[[1, 2],
[3, 4]]
```
This should probably be specified in the docstring.

**Comment by adityaranjan** [comment]
> Which dimensions are the senders/receivers?
> 
> For example in this matrix:
> 
> ```
> [[1, 2],
> [3, 4]]
> ```
> 
> This should probably be specified in the docstring.

Yeah, it is specified in a small comment. I'll put it in the top-level one.

**Review by hsirkar — APPROVED** [review]
Looks good.

File: pipit/trace.py [hunk 1]
```diff
@@ -29,3 +31,77 @@ def from_hpctoolkit(dirname):
         from .readers.hpctoolkit_reader import HPCToolkitReader
 
         return HPCToolkitReader(dirname).read()
+
+    def comm_matrix(self, comm_type="bytes"):
```
Comment (bhatele) [inline pipit/trace.py:35]: we should change comm_type to output and the values we should allow are count or size

File: pipit/trace.py [hunk 2]
```diff
@@ -29,3 +31,77 @@ def from_hpctoolkit(dirname):
...
+        receiver_ranks = (
+            sender_dataframe["Attributes"]
+            .apply(lambda attrDict: attrDict["receiver"])
+            .to_list()
+        )
+
+        # number of bytes communicated
+        if comm_type == "bytes":
+            # (1 communication is a single row in the sender dataframe)
+            message_volumes = (
```
Comment (bhatele) [inline pipit/trace.py:87]: the plural of volume is also plural

**Pull request merged**